### PR TITLE
deep merge fragments in order to avoid dropping keys

### PR DESCRIPTION
**Pull request by charlieschwabacher**
resolves https://github.com/rmosolgo/graphql-ruby/issues/29

### CODE REVIEW

File: lib/graphql/query/serial_execution/selection_resolution.rb [hunk 1]
```diff
@@ -21,12 +21,29 @@ def initialize(target, type, selections, query, execution_strategy)
...
+            deep_merge memo, field_value
           end
         end
 
         private
 
+        def deep_merge(h1, h2)
+          h1.merge(h2) do |key, oldval, newval|
+            oldval = oldval.to_hash if oldval.respond_to?(:to_hash)
+            newval = newval.to_hash if newval.respond_to?(:to_hash)
```
Comment (rmosolgo) [inline lib/graphql/query/serial_execution/selection_resolution.rb]: Why is this check / coercion necessary?

File: lib/graphql/query/serial_execution/selection_resolution.rb [hunk 2]
```diff
@@ -21,12 +21,29 @@ def initialize(target, type, selections, query, execution_strategy)
...
+
+            if oldval.class.to_s == 'Array' && newval.class.to_s == 'Array'
+              oldval.each_index.map do |i|
+                deep_merge oldval[i], newval[i]
+              end
+            elsif oldval.class.to_s == 'Hash' && newval.class.to_s == 'Hash'
```
Comment (rmosolgo) [inline lib/graphql/query/serial_execution/selection_resolution.rb]: Is there an advantage to `class.to_s` over `.is_a?` ?

**Comment by rmosolgo** [comment]
Cool, the direction of this implementation seems good, how about adding a test? 

You could put it in query_spec.rb with a new `describe "merging fragments with different keys" do ...` block. You could set up a `let(:query_string)` with a query with 2 fragments, then make sure the result has fields from both fragments.

**Comment by charlieschwabacher** [comment]
I had just pulled that deep merge method from ActiveSupport, you are right that the coercion isn't needed / is_a? is faster.

**Comment by rmosolgo** [comment]
:money_with_wings: awesome, thanks!

**Comment by rmosolgo** [comment]
just pushed 0.7.1 with this fix!

**Pull request merged**